gneiss/balances.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import division
+import numpy as np
+from skbio.stats.composition import clr_inv
+from collections import OrderedDict
+
+
+def _balance_basis(tree_node):
+    """ Helper method for calculating balance basis
+    """
+
+    n_tips = sum([n.is_tip() for n in tree_node.traverse()])
+    counts = _count_matrix(tree_node)
+    counts = OrderedDict([(x, counts[x])
+                          for x in counts.keys() if not x.is_tip()])
+    nds = counts.keys()
+    r = np.array([counts[n]['r'] for n in nds])
+    s = np.array([counts[n]['l'] for n in nds])
+    k = np.array([counts[n]['k'] for n in nds])
+    t = np.array([counts[n]['t'] for n in nds])
+
+    a = np.sqrt(s / (r*(r+s)))
+    b = -1*np.sqrt(r / (s*(r+s)))
+
+    basis = np.zeros((n_tips-1, n_tips))
+    for i in range(len(nds)):
+        basis[i, :] = np.array([0]*k[i] + [a[i]]*r[i] + [b[i]]*s[i] + [0]*t[i])
+    return basis, nds
+
+
+def balance_basis(tree_node):
+    """
+    Determines the basis based on tree
+
+    Parameters
+    ----------
+    treenode : skbio.TreeNode
+        Phylogenetic tree.  Must be a strictly bifurcating tree
+    Returns
+    -------
+    basis : np.array
+        Returns a set of orthonormal bases in the Aitchison simplex
+        corresponding to the phylogenetic tree. The order of the
+        basis is index by the level order of the internal nodes.
+    nodes : list, skbio.TreeNode
+        List of tree nodes indicating the ordering in the basis.
+
+    Raises
+    ------
+    ValueError
+        The tree doesn't contain two branches
+
+    Examples
+    --------
+    >>> from gneiss.balances import balance_basis
+    >>> from skbio import TreeNode
+    >>> tree = u"((b,c)a, d)root;"
+    >>> t = TreeNode.read([tree])
+    >>> basis, nodes = balance_basis(t)
+    >>> basis
+    array([[ 0.62985567,  0.18507216,  0.18507216],
+           [ 0.28399541,  0.57597535,  0.14002925]])
+
+    Notes
+    -----
+    The tree must be strictly bifurcating, meaning that
+    every internal node has exactly 2 children.
+    """
+    basis, nodes = _balance_basis(tree_node)
+    basis = clr_inv(basis)
+    return basis, nodes
+
+
+def _count_matrix(treenode):
+    nodes = list(treenode.levelorder(include_self=True))
+    # fill in the dictionary
+    counts = OrderedDict()
+    columns = ['k', 'r', 'l', 't', 'tips']
+    for n in nodes:
```

Comment:
Thanks for linking!  If this were a data structure that will be continuously used, I'd think about it.

I don't think it is necessary in this context, since this `OrderedDict` is only used to store intermediate results.  The only thing that matters is the matrix, and the nodes that are returned.